File: 8_convert_console_ui/src/main/java/dojo/ConsoleUI.java
package dojo;

import java.io.IOException;
import java.util.function.Consumer;

public class ConsoleUI {
    private final FunctionThatThrows<String> in;
    private final Consumer<String> out;
    private Game game;

    public ConsoleUI(Game game, FunctionThatThrows<String> in, Consumer<String> out) {
        this.game = game;
        this.in = in;
        this.out = out;
    }

    public void run() {
        out.accept(render());

        while (!game.isGameOver()) {
            try {
                String input = in.apply();
                if (input == null)
                    break;
                Move move = parseInput(input);
                game = game.makeMove(move, game.nextCellToPlace());
                out.accept(render());
            } catch (Game.InvalidMove invalidMove) {
                out.accept(invalidMove.getMessage() + "\n" + render());
            } catch (IOException e) {
                out.accept("Unrecoverable error");
            }
        }
    }

    public String render() {
        return new GameRenderer(game).render();
    }

    private Move parseInput(String input) throws Game.InvalidMove {
        try {
            String[] split = input.split(" ");
            if (split.length < 2) {
                throw new Game.InvalidMove(invalidInputMessage(input));
            }
            int row = Integer.parseInt(split[0]);
            int column = Integer.parseInt(split[1]);
            return new Move(row, column);
        } catch (NumberFormatException e) {
            throw new Game.InvalidMove(invalidInputMessage(input));
        }
    }

    private String invalidInputMessage(String s) {
        return "Invalid input '" + s + "', please make sure it's zero-indexed 'x y' coordinates";
    }
}

File: 11_final_clean_up/src/main/kotlin/dojo/ConsoleUI.kt
package dojo

import java.io.IOException
import java.util.function.Consumer

class ConsoleUI(private var game: Game, private val `in`: FunctionThatThrows<String?>, private val out: Consumer<String>) {

    fun run() {
        out.accept(render())

        while (!game.isGameOver) {
            try {
                val input = `in`.apply() ?: break
                val move = parseInput(input)
                game = game.makeMove(move, game.nextCellToPlace())
                out.accept(render())
            } catch (invalidMove: Game.InvalidMove) {
                out.accept(invalidMove.message + "\n" + render())
            } catch (e: IOException) {
                out.accept("Unrecoverable error")
            }

        }
    }

    fun render() = GameRenderer(game).render()

    private fun parseInput(input: String): Move {
        try {
            val split = input.split(" ")
            if (split.size < 2) {
                throw Game.InvalidMove(invalidInputMessage(input))
            }
            val row = Integer.parseInt(split[0])
            val column = Integer.parseInt(split[1])
            return Move(row, column)
        } catch (e: NumberFormatException) {
            throw Game.InvalidMove(invalidInputMessage(input))
        }

    }

    private fun invalidInputMessage(s: String) = "Invalid input '$s', please make sure it's zero-indexed 'x y' coordinates"
}
